ROMaterials: Skin/TPS selection inside Editor

fluff-li opened the pull request:
First, I'm not sure if this is the right Repo for this, it started as an extension to ROHeatshields but grew bigger.

This is a material based thermal property extension, inspired by [RO_Materials.cfg](https://github.com/KSP-RO/RealismOverhaul/blob/master/GameData/RealismOverhaul/RO_Materials.cfg) & [ROHeatshields](https://github.com/KSP-RO/ROHeatshields), adding the ability to select the Skin/Thermal Protection System of a part inside Editor.

![Screenshot (233)](https://github.com/KSP-RO/ROLibrary/assets/56023678/04a8db2f-179c-468d-aea5-7cbac848c53c)
The selected preset effects all skin related thermal properties, mass and cost.
The preset list is defined per part base. The preset definitions are not attached to parts.

![Screenshot (234)](https://github.com/KSP-RO/ROLibrary/assets/56023678/f04f9e99-84f7-497e-8268-0ae00985aabf)
Core material changes based on tank types from Realfuels and in connection also the part properties.

Variables "thermalMassModifier" (Specific Heat Capacity), "skinInternalConductionMult" (Thermal Conductivity) and "emissiveConstant" change based on skin temperature, mimicking real-life behavior.

Tool for creating the Preset configs and csv tables [fluff-li/RoMaterials_csv_calculator](https://github.com/fluff-li/RoMaterials_csv_calculator)
  
  
  
~~Depends on: [ProceduralParts - fork:](https://github.com/fluff-li/ProceduralParts) to calculate surface area on procedural parts~~

To Do:
-   Add cost values & balancing
-   ~~Automatic Reentry Tagging in ModuleTagList~~
-   ~~Check interaction with MLI layering~~ 
    - Parts with MLI layer directly change part.heatConductivity based on partPrefab.skinInternalConductionMult
    - Other calculations are also partially based values from partPrefab
-   Presets: expand and revise

WIP/Bugs:
-   ~~Unreliable/wrong surface area calculation using FARAeroPartModule, editor & inflight values differ~~
       ~~- The big inconsistencies seem to occur only on Procedural & ModularTank parts~~
-   ~~ModuleROTank: changing Nose/Mount variant doesn't update data~~
-   ~~GUI: part weight lags behind, when tank type gets changed~~
-   ~~Surface attached parts/children don't calculate contact area~~ Gets taken care of by FAR voxelization

fluff-li committed:
ROMaterials Module: change thermal prop. based on selected material

Parts implementing this, get a customization option inside VAB/SPH, for both core & exterior shell material/construction, which in turn affects the thermal properties of the part as a whole.

fluff-li committed:
cleaned code comments, added Preset & PartUpgrade cfg + Notes

fluff-li committed:
ModuleROMaterials.cs & PresetROMaterials.cfg

PresetROMaterials.cfg: add variable explanation, cleaned names
ModuleROMaterials.cs:
  - make core choose option inactive by default
  - save core & skin values defined in cfg
  - improve implementation of core preset defined in cfg
  - clean how presetCoreName gets overwriten, make it go through PresetCore
  - move thermal mass calculation in updateGUI function
  - make updateHight function differentiate new skin preset & not

fluff-li committed:
-Added:
Flight display to GUI with basic information
SkinHeight option for Part in cfg Module
Function to make shure core/skin defined in cfg gets addet to core-/skinPresets list
Nullref Exception check for UpdateCoreForRealfuels()

-Expanded
ThermalConductivity calculation
ApplySkinPreset() function to foreward skinNew parameter to UpdateHeight()

Handle skin to skin conductivity value same as skin to internal for now

Expanded Matarials.txt

Revised PresetsROMaterial.cfg explanation on start of the file

NathanKell commented:
HERO. :)

fluff-li commented:
I forgot to mention, it is still WIP, the configs are in early stages and the code has some bugs. I'm going append a list of features/bugs tagged with wip, missing, planned & finished.

Capkirk123 commented:
Hot damn! That's good.

NovemberOrWhatever commented:
That's amazing, well done

fluff-li committed:
Added support for a list of attach nodes to be ignored when subtracting contact area from surface area and also the option to ignore all surface attached parts, for the same thing. The main idea is to use this on parts that can hold internal attachments, like corgo bays.

fluff-li committed:
more work on dyna-soar preset. Detailed information is scarce unlike the shuttle.
added an excel sheet to ease preset calculation
moved temporary part patches for testing in here

fluff-li committed:
fixed excel sheet formula mistake & slightly improved accessibility

fluff-li committed:
heat capacity calculation overhaul

Added a multiplayer to heat capacity of all surface sections & layer, based on the fraction between max temp of section/layer & the part/section. This is to compensate the dissonance between the supposed temperature difference between layers & section & kerbals uniform temperature modelling.

fluff-li committed:
Fix NullReferenceException for Presets that are no longer available, eg. when loading a craft with removed or renamed preset.

fluff-li committed:
Fix NullRef in caused by UpdateCoreForRealfuels();

fluff-li committed:
Don't override part.termalmass & -skinthermalmass on Gui Update

fluff-li committed:
fixed mistakes in set_PresetCore, made sure ApplyCorepreset doesn't get called twice om OmStertFinished

fluff-li committed:
tweaked Dyna-Soar layering configuration and added Space Shuttle to excel sheet

fluff-li committed:
- Implemented a workaround to wait for FARs voxelizaion when reading surface area
- Default to fARAeroPartModule for surface area calculation
- Removed no longer needed debug logs
- Implemented a calculation to keep the thermal mass the same on TPS change

fluff-li committed:
Let Editor know that mass has changed by fireing onEditorShipModified when TPS height changes & skip waiting for voxelization

NathanKell commented:
Sorry, for some reason our CI is trying to build this despite it being on another fork. Please ignore any spam from github in that regard. :)

fluff-li committed:
Fix mass diplay lagging behind & more

Build a workaround to update mass diplay in GUI
Currently, the modules ModuleFuelTanks and FAR update their mass through periodic checks in the Update() / FixedUpdate() functions, which causes a delay in reflecting changes made through the uiControlEditor. As a result, the onFieldChanged calls on these modules execute before they update their mass.

Other Changes:
- Use OnPartVolumeChanged event to catch volume changes in ModuleROTanks
- Fire EditorShipModified event sender only if the module mass has changed

fluff-li committed:
-Use average heat capacity instead of the value on temperature limit
-Add eraly shuttle & shuttle presets
-Tweak excel sheet calculations

fluff-li commented:
> Sorry, for some reason our CI is trying to build this despite it being on another fork. Please ignore any spam from github in that regard. :)

No worries :)

fluff-li committed:
Implemented automatic reentry tagging
- reduced number of tries of rechecking for mass update
- ignore next voxelization if onEditorShipModified is fierd from ROMaterials

fluff-li committed:
Experimenting on presets based on properties,
where the value of skinMaxTemp is set to average temperature across part, instead of localised maximum.

fluff-li committed:
Added functionality to incorporate temperature-dependent thermal property values

Thermal mass value of the core is no longer influenced by thermal mass of skin and vice versa

fluff-li committed:
implemented a minimum check for core mass to prevent negative thermal mass value, after subtracting skin mass

fluff-li committed:
Implement min & max value handling for temperature dependent properties.

Load temperature dependent properties array into Module on Loading if LoadedSceneIsFlight.

Removed part.heatConductivity from skinInternalConductionMult calculation.

Check for infinity on heightfactor calculation after TPS change.

fluff-li committed:
Changed skinHeightMin/Max to save in mm instead of m

fluff-li committed:
Updating Preset cfg & csv, created with github.com/fluff-li/RoMaterials_csv_calculator

Updated Notes

fluff-li committed:
Handle another NaN value for heightFactor, which happens when skinHeightMax == skinHeightMin

fluff-li committed:
implemented temperature dependend porperties for core, rename presetTPS to presetSkin

fluff-li committed:
Removed duplicate SkinInternalConductivityDivGlobal multiplier

fluff-li committed:
Implemented PAW open & "Display Thermal Data in Action Menus" toggle check for displaying & updating Flight Debug Display

fluff-li committed:
Add PhysicsGlobals.ThermalDataDisplay check on guiActivator

fluff-li committed:
clean up

fluff-li committed:
Removed 0.5 multiplier from SkinInternalConductivityDivGlobl, which effects skinInternalConductionMult calculation

fluff-li committed:
- give priority to manualy enterd part surface area over far calculation,
- csv directory reorganization

fluff-li committed:
Switched multidimensional arrays with more performant jagged array

fluff-li committed:
Added exception handling and improved reading & saving of csv files

Removed persistance of FlightDisplay & Debug Fileds

Temporarliy disabled getting surface area from ProceduralParts module

Split LoadThermalPropertiesAttay function into two for skin & core

Small changes on Debug.Logs

fluff-li committed:
Modified *DivGlobal variales to get methods & moved heatConductivity manipulation to ApplyCorePreset

fluff-li committed:
Unified all (PhysicsGlobals.StandardSpecificHeatCapacity * part.thermalMassModifier) multiplier into one get Method

fluff-li committed:
Fixed Nullref exceptions caused by using & instead of &&

fluff-li committed:
Removed every Method of obtaining surface area besides of far & direct definition via cfg, scince with far installed, the game gets its in flight surface area from far

Fixed bug with flightDebug display

fluff-li committed:
Switched heatConductivityDivGlobal with heatConductivityDivGlobalSqrt, adjusting it to FlightIntegrator calculation

fluff-li committed:
Restrucurted GameData folder & added new presets

fluff-li committed:
Update .csproj

fluff-li committed:
Merge branch 'KSP-RO:master' into master

fluff-li committed:
Replaced workaound for checking of far voxelization status with KSPEvent fired via harmony patch

fluff-li committed:
Overhauled SetSurfaceArea function as UpdateSurfaceArea & handling of Height change for Skin

fluff-li committed:
Removed obsolete Editor EditorCordinator.cs from csproj

fluff-li committed:
Changed workings of flight debug display on part window & added entry to input a referance surface density for module mass calculation

fluff-li committed:
Expanded zero point entries to include preset reference & cost per area

fluff-li committed:
Implemented switching between UI_ChooseOption & alternative displaying of core preset, based on value defined in config

fluff-li committed:
Implemented coreThermalMassFraction read from config; fixed bug in UpdateGUI

fluff-li committed:
Overhauled Harmony patch for UpdateMassStats

fluff-li committed:
Fixed core preset & skin height not persisting

fluff-li committed:
Increased significant figures of skin height slider

fluff-li committed:
Updated GameData/ & Notes/

NathanKell commented:
@fluff-li please let us know when you're ready for this to be tested / merged! :)
